crates/wasmtime/src/engine.rs

```diff
@@ -55,6 +55,11 @@ struct EngineInner {
 impl Engine {
     /// Creates a new [`Engine`] with the specified compilation and
     /// configuration settings.
+    ///
+    /// # Errors
+    ///
+    /// This method can fail if the `config` is invalid or some
+    /// configurations are incompatible.
```

Comment:
Agreed about expanding the error documentation a bit here, but I also think it would be best to keep documentation on the original config methods as well. For example `Config::cranelift_flag_set` could retain some documentation indicating that if an invalid setting is specified that the error will pop out during `Engine::new` here.

Reply:
> Can we say a bit more here? Something about how some Wasm features require certain compiler features, and they will turn on the relevant compiler features by default, but explicitly disabling some compiler features or setting them in an incompatible way may cause errors.

Fixed.

> I think we should also note a guarantee/promise that default compiler options (i.e., if the user does not add any specific setting overrides) will always work if the underlying hardware supports what is needed.

I don't think it necessary to add this note explicitly, because the default options should work is intuition for users.

Reply:
On https://docs.wasmtime.dev/c-api/config_8h.html#ada735dd95cbee6854fc6bb27dc9591c6 it says:

> If the compilation strategy selected could not be enabled then an error is returned. 

I've been looking everywhere in the docs to try and find what happened to config errors, I don't think it is very clear today without looking into this PR.

Reply:
Hi, thanks for your remind.
I changed the doc and the signature of the rust code, but I forgot the doc for c.
I've submitted #4550 to fix this.

Reply:
@PureWhiteWu That's great! Very prompt, thank you!

A question still remains on my end but I am not sure if here's the right place for it, I'll shoot it again though:

Since `wasm.h`'s `wasm_engine_new_with_config` does not return any error, where are the engine initialization errors reported when using the C API? 🤔

Reply:
Sorry, I'm not familiar with the c api, @alexcrichton do you know about this?

Reply:
You can [take a look at the implementation](https://github.com/bytecodealliance/wasmtime/blob/fb7d51033c4d507f412d7f5ec580cb325d5a4e0a/crates/c-api/src/engine.rs#L29-L35) and right now it looks like no error handling happens and if something otherwise would result in an error it would cause the process to abort.

This is probably something that would be good to fix, returning a null pointer there and perhaps adding a Wasmtime-specific API for the same operation but one that returns an error message.